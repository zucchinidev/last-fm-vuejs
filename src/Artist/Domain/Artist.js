export class Artist {
  constructor ({ name, url, associateImages: images } = {}) {
    this.name = name
    this.url = url
    this.images = images
  }
}
