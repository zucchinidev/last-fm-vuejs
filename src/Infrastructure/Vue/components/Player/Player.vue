<template>
  <v-flex xs12 sm6 offset-sm3>
    <v-card v-if="hasTrack()" class="white--text">
      <v-container>
        <v-layout row>
          <v-flex xs7>
            <v-card-media
              :src="selectedTrack.getSmallImage().url"
              height="125px"
              contain
            ></v-card-media>
          </v-flex>
          <v-flex xs5>
            <div>
              <div class="headline">
                {{selectedTrack.name}}
              </div>
              <div>
                {{selectedTrack.duration | ms-to-mm}}
              </div>
            </div>
            <v-card-actions>
              <v-spacer></v-spacer>
              <audio :src="selectedTrack.getPreviewUrl()" controls></audio>
              <v-spacer></v-spacer>
            </v-card-actions>
          </v-flex>
        </v-layout>
      </v-container>
    </v-card>
  </v-flex>
</template>
<script>
export default {
  name: 'Player',
  props: {
    /** @type Track */
    selectedTrack: {
      type: Object
    }
  },
  methods: {
    hasTrack () {
      return this.selectedTrack && this.selectedTrack.album
    }
  },
  created () {
    this.$bus.$on('set-track', track => (this.selectedTrack = track))
  }
}
</script>
<style scoped>
  .card__media {
    clip-path: circle(30%);
  }
</style>
