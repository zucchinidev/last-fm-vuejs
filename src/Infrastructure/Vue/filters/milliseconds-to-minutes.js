export const calculateResult = (originalValue) => {
  const milliseconds = Number(originalValue)
  const minutes = Math.floor(milliseconds / 60000)
  const seconds = (minutes % 60000 / 1000).toFixed(0)
  let result = minutes < 10 ? `0${minutes}:` : `${minutes}:`
  result += seconds < 10 ? `0${seconds}` : seconds
  return result
}
export const millisecondsToMinutes = {
  install (Vue) {
    Vue.filter('ms-to-mm', calculateResult)
  }
}
