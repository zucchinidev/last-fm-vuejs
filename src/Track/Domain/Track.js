const availableSizes = {
  small: 64,
  medium: 300,
  big: 640
}

export class Track {
  constructor ({ album, duration, name, popularity, images, id, previewUrl } = {}) {
    this.id = id
    this.album = album
    this.duration = duration
    this.name = name
    this.popularity = popularity
    this.images = images
    this.previewUrl = previewUrl
  }

  getSmallImage () {
    return this.images.find(image => image.width === availableSizes.small)
  }

  getBigImage () {
    return this.images.find(image => image.width === availableSizes.big)
  }

  getPreviewUrl () {
    return this.previewUrl
  }
}
