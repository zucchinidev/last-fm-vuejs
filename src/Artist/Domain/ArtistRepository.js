import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import config from '../../Infrastructure/lastFmApi/config'
import { Artist } from './Artist'
const urls = {
  getArtistsByCountry: ''
}

const configuration = {
  slug: 'contacts',
  baseURL: 'http://ws.audioscrobbler.com/2.0/'
}
export const ArtistRepository = {
  async getTopArtistsByCountry (country = 'spain') {
    const params = {
      method: 'geo.gettopartists',
      country,
      api_key: config.apiKey,
      format: 'json'
    }
    const { data } = await HttpRepository.get(urls.getArtistsByCountry, params, configuration)
    return data.topartists.artist.map(artist => new Artist({
      images: artist.image.map(image => ({url: image['#text'], size: image.size})),
      url: artist.url,
      name: artist.name
    }))
  }
}
