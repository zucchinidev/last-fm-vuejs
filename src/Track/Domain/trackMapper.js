import { Track } from './Track'

export default (data) => {
  const artists = (data.album.artists || []).map(artist => artist.name)
  const images = data.album.images || {}
  return new Track({
    album: {
      artists,
      name: data.album.name,
      release: data.album.release_date
    },
    images,
    duration: data.duration_ms,
    name: data.name,
    popularity: data.popularity
  })
}
