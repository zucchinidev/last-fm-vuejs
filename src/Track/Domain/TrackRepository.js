import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import trackMapper from './trackMapper'

const urls = {
  search: 'search'
}

const configuration = {
  baseURL: process.env.API_MUSIC_BASE_URL
}
export const TrackRepository = {
  async searchTrack (searchTerm) {
    const params = {
      q: searchTerm,
      type: 'track'
    }
    const { data } = await HttpRepository.get(urls.search, params, configuration)
    return data.tracks.items.map(trackMapper)
  }
}
