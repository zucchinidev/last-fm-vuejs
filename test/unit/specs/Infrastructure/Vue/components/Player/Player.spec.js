import Player from '@Vue/components/Player/Player'
import { WrapComponent } from '../../../../../helpers'
import { Track } from '../../../../../../../src/Track/Domain/Track'
import * as fixture from '../../../../../fixtures/trackFixture'
import PlayerPageObject from './PlayerPageObject'
import { calculateResult } from '../../../../../../../src/Infrastructure/Vue/filters/milliseconds-to-minutes'

const { track: trackFixture } = fixture
describe('Player.vue', () => {
  let wrapper
  /** @type Track */
  let track
  /** @type PlayerPageObject */
  let page
  beforeEach(() => {
    track = new Track(trackFixture)
    wrapper = WrapComponent(Player).withProps({ track }).mount()
    page = new PlayerPageObject(wrapper)
  })
  it('should render correct contents', async () => {
    await page.wait()
    expect(page.getTrackName()).toEqual(trackFixture.name)
    expect(page.getDuration()).toEqual(calculateResult(trackFixture.duration))
    expect(page.getPoster()).toEqual(track.getSmallImage().url)
    expect(page.getAudio()).toEqual(track.getPreviewUrl())
  })

  it('must receive a track as property', () => {
    expect(wrapper.props().track).toEqual(track)
  })

  it('should has correct html', function () {
    page.matchSnapshot()
  })
})
