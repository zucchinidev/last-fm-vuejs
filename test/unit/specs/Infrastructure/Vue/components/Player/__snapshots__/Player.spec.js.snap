// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player.vue should has correct html 1`] = `
<div class="container fluid">
    <div class="layout row">
        <div class="flex xs3">
            <div class="track__data__container">
                <img src="https://i.scdn.co/image/31dd0b4e46c5fb0feb1e02c2f98d07e76ca5ab6c" alt="Esa Muchacha">
                <div>
                    <p class="title">
                        Esa Muchacha
                    </p>
                    <p>
                        Esa Muchacha
                        <span>03:00</span></p>
                </div>
            </div>
        </div>
        <div class="flex xs9">
            <div class="audio__container">
                <audio src="https://p.scdn.co/mp3-preview/a1db7be86b3d256d2e1bf562460947722c1aff01?cid=555776939cf64ea6b39915cf4d5d875d" controls="controls"></audio>
            </div>
        </div>
    </div>
</div>
`;
